test_data_too_large failing on i686-linux since 42.0.0

We're seeing various test failures in nixpkgs due to the following error on i686-linux.

```
    def large_mmap():
>       return mmap.mmap(-1, 2**32, prot=mmap.PROT_READ)
E       OverflowError: Python int too large to convert to C ssize_t
```

This was likely introduced in https://github.com/pyca/cryptography/commit/596e8fbb480f5278dbbccc221fb18bb8b03b802c.

Is 32 Bit support something you are still looking into? This affects e.g. Lutris for us as a distro.

### Environment
- Python 3.11.7
- cryptography 42.0.2
- OpenSSL 3.0.13
- cffi 1.16.0
- setuptools 69.0.3
- pip 23.3.1

### Test log

```
cryptography-i686-linux> ________________________ TestAESCCM.test_data_too_large ________________________
cryptography-i686-linux> [gw12] linux -- Python 3.11.7 /nix/store/ah5382l41gaffmx8yj8w3j34n18mh1al-python3-3.11.7/bin/python3.11
cryptography-i686-linux> 
cryptography-i686-linux> self = <tests.hazmat.primitives.test_aead.TestAESCCM object at 0xf55df8f0>
cryptography-i686-linux> 
cryptography-i686-linux>     @pytest.mark.skipif(
cryptography-i686-linux>         sys.platform not in {"linux", "darwin"}, reason="mmap required"
cryptography-i686-linux>     )
cryptography-i686-linux>     def test_data_too_large(self):
cryptography-i686-linux>         key = AESCCM.generate_key(128)
cryptography-i686-linux>         aesccm = AESCCM(key)
cryptography-i686-linux>         nonce = b"0" * 12
cryptography-i686-linux>     
cryptography-i686-linux> >       large_data = large_mmap()
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:207: 
cryptography-i686-linux> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cryptography-i686-linux> 
cryptography-i686-linux>     def large_mmap():
cryptography-i686-linux> >       return mmap.mmap(-1, 2**32, prot=mmap.PROT_READ)
cryptography-i686-linux> E       OverflowError: Python int too large to convert to C ssize_t
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:41: OverflowError
cryptography-i686-linux> ___________________ TestChaCha20Poly1305.test_data_too_large ___________________
cryptography-i686-linux> [gw11] linux -- Python 3.11.7 /nix/store/ah5382l41gaffmx8yj8w3j34n18mh1al-python3-3.11.7/bin/python3.11
cryptography-i686-linux> 
cryptography-i686-linux> self = <tests.hazmat.primitives.test_aead.TestChaCha20Poly1305 object at 0xf55dc2d0>
cryptography-i686-linux> 
cryptography-i686-linux>     @pytest.mark.skipif(
cryptography-i686-linux>         sys.platform not in {"linux", "darwin"}, reason="mmap required"
cryptography-i686-linux>     )
cryptography-i686-linux>     def test_data_too_large(self):
cryptography-i686-linux>         key = ChaCha20Poly1305.generate_key()
cryptography-i686-linux>         chacha = ChaCha20Poly1305(key)
cryptography-i686-linux>         nonce = b"0" * 12
cryptography-i686-linux>     
cryptography-i686-linux> >       large_data = large_mmap()
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:66: 
cryptography-i686-linux> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cryptography-i686-linux> 
cryptography-i686-linux>     def large_mmap():
cryptography-i686-linux> >       return mmap.mmap(-1, 2**32, prot=mmap.PROT_READ)
cryptography-i686-linux> E       OverflowError: Python int too large to convert to C ssize_t
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:41: OverflowError
cryptography-i686-linux> ________________________ TestAESGCM.test_data_too_large ________________________
cryptography-i686-linux> [gw13] linux -- Python 3.11.7 /nix/store/ah5382l41gaffmx8yj8w3j34n18mh1al-python3-3.11.7/bin/python3.11
cryptography-i686-linux> 
cryptography-i686-linux> self = <tests.hazmat.primitives.test_aead.TestAESGCM object at 0xf55e3d10>
cryptography-i686-linux> 
cryptography-i686-linux>     @pytest.mark.skipif(
cryptography-i686-linux>         sys.platform not in {"linux", "darwin"}, reason="mmap required"
cryptography-i686-linux>     )
cryptography-i686-linux>     def test_data_too_large(self):
cryptography-i686-linux>         key = AESGCM.generate_key(128)
cryptography-i686-linux>         aesgcm = AESGCM(key)
cryptography-i686-linux>         nonce = b"0" * 12
cryptography-i686-linux>     
cryptography-i686-linux> >       large_data = large_mmap()
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:385: 
cryptography-i686-linux> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cryptography-i686-linux> 
cryptography-i686-linux>     def large_mmap():
cryptography-i686-linux> >       return mmap.mmap(-1, 2**32, prot=mmap.PROT_READ)
cryptography-i686-linux> E       OverflowError: Python int too large to convert to C ssize_t
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:41: OverflowError
cryptography-i686-linux> _______________________ TestAESOCB3.test_data_too_large ________________________
cryptography-i686-linux> [gw14] linux -- Python 3.11.7 /nix/store/ah5382l41gaffmx8yj8w3j34n18mh1al-python3-3.11.7/bin/python3.11
cryptography-i686-linux> 
cryptography-i686-linux> self = <tests.hazmat.primitives.test_aead.TestAESOCB3 object at 0xf55eb370>
cryptography-i686-linux> 
cryptography-i686-linux>     @pytest.mark.skipif(
cryptography-i686-linux>         sys.platform not in {"linux", "darwin"}, reason="mmap required"
cryptography-i686-linux>     )
cryptography-i686-linux>     def test_data_too_large(self):
cryptography-i686-linux>         key = AESOCB3.generate_key(128)
cryptography-i686-linux>         aesocb3 = AESOCB3(key)
cryptography-i686-linux>         nonce = b"0" * 12
cryptography-i686-linux>     
cryptography-i686-linux> >       large_data = large_mmap()
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:530: 
cryptography-i686-linux> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cryptography-i686-linux> 
cryptography-i686-linux>     def large_mmap():
cryptography-i686-linux> >       return mmap.mmap(-1, 2**32, prot=mmap.PROT_READ)
cryptography-i686-linux> E       OverflowError: Python int too large to convert to C ssize_t
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:41: OverflowError
cryptography-i686-linux> ________________________ TestAESSIV.test_data_too_large ________________________
cryptography-i686-linux> [gw15] linux -- Python 3.11.7 /nix/store/ah5382l41gaffmx8yj8w3j34n18mh1al-python3-3.11.7/bin/python3.11
cryptography-i686-linux> 
cryptography-i686-linux> self = <tests.hazmat.primitives.test_aead.TestAESSIV object at 0xf5571470>
cryptography-i686-linux> 
cryptography-i686-linux>     @pytest.mark.skipif(
cryptography-i686-linux>         sys.platform not in {"linux", "darwin"}, reason="mmap required"
cryptography-i686-linux>     )
cryptography-i686-linux>     def test_data_too_large(self):
cryptography-i686-linux>         key = AESSIV.generate_key(256)
cryptography-i686-linux>         aessiv = AESSIV(key)
cryptography-i686-linux>     
cryptography-i686-linux> >       large_data = large_mmap()
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:704: 
cryptography-i686-linux> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cryptography-i686-linux> 
cryptography-i686-linux>     def large_mmap():
cryptography-i686-linux> >       return mmap.mmap(-1, 2**32, prot=mmap.PROT_READ)
cryptography-i686-linux> E       OverflowError: Python int too large to convert to C ssize_t
cryptography-i686-linux> 
cryptography-i686-linux> tests/hazmat/primitives/test_aead.py:41: OverflowError
```

--- a/tests/hazmat/primitives/test_aead.py
+++ b/tests/hazmat/primitives/test_aead.py
@@ -38,7 +38,11 @@ def _aead_supported(cls):
 
 
 def large_mmap():
-    return mmap.mmap(-1, 2**32, prot=mmap.PROT_READ)
+    # We need this large but not larger than fits in a 32-bit int. This way
+    # a 32-bit platform can return this mmap successfully but we'll raise
+    # OverFlowError in the tests because the underlying type for the
+    # function signature is a signed int
+    return mmap.mmap(-1, 2**31, prot=mmap.PROT_READ)
 
 
 @pytest.mark.skipif(

